### PR TITLE
feat: unify lib-uncommons baseline into lib-commons v4

### PR DESCRIPTION
**Pull request by fredcamaral**
## Summary

Complete unification of the `lib-uncommons` baseline into `lib-commons` v4. This is a ground-up modernization of the shared library, introducing explicit error returns, nil-safety, structured telemetry, and new resilience primitives across all packages.

### Infrastructure & Build
- **Module path**: migrated to `github.com/LerianStudio/lib-commons/v4`
- **Build system**: modernized Makefile with unit/integration/coverage targets, `LOW_RESOURCE` and `RETRY_ON_FAIL` flags
- **Linting & release**: updated `.golangci.yml` and `.goreleaser.yml` for v4
- **Dependencies**: OTEL 1.42.0, gRPC 1.79.2, and all patch-level bumps

### Core Rewrite
- **Logger interface** (`commons/log`): 5-method `Logger` contract with structured `Field` constructors, CWE-117 log-injection prevention, and `GoLogger` stdlib adapter
- **Zap adapter** (`commons/zap`): OTEL bridge support, typed `Config` with environment-aware defaults
- **Constants**: expanded OTEL attributes, metric label sanitization, transaction lifecycle constants

### New Packages
- **`commons/assert`**: production-safe assertions with telemetry integration and domain predicates (UUID, amount, balance, port validation)
- **`commons/backoff`**: exponential backoff with jitter and context-aware sleep
- **`commons/runtime`**: panic recovery policies, safe goroutine wrappers, panic metrics, error reporter, production mode
- **`commons/safe`**: panic-free math (divide, percentage), cached regex, and slice operations with error returns
- **`commons/errgroup`**: goroutine coordination with panic recovery
- **`commons/cron`**: cron expression parsing and scheduling
- **`commons/jwt`**: HMAC-based JWT signing, verification, and time-claim validation
- **`commons/outbox`**: transactional outbox pattern with Postgres repository, tenant isolation, event classification, and dispatcher

### Refactored Packages
- **Telemetry** (`commons/opentelemetry`): `Redactor` with `RedactionRule` patterns replaces `FieldObfuscator`, `RedactingAttrBagSpanProcessor`, explicit `TelemetryConfig`
- **Metrics** (`commons/opentelemetry/metrics`): `MetricsFactory` + fluent builders (Counter, Gauge, Histogram), `NopFactory` for tests
- **HTTP** (`commons/net/http`): consolidated response/error rendering, cursor/offset/sort pagination, `ParseBodyAndValidate`, SSRF-protected reverse proxy, tenant/resource ownership verification, rate limit storage
- **Server** (`commons/server`): `ServerManager` lifecycle with `WithHTTPServer`/`WithGRPCServer`/`WithShutdownHook` chainable config
- **Circuit breaker**: preset configs (Default, Aggressive, Conservative, HTTP, Database), metrics integration, validated `Manager`
- **Transaction**: intent-based planning (`BuildIntentPlan`), balance eligibility validation, posting flow with `ResolveOperation`
- **Data connectors**: Postgres (dbresolver, lazy-connect with backoff, migrations), MongoDB (functional options, URI builder, index helpers), Redis (topology-based config, GCP IAM auth, Redsync locking), RabbitMQ (context-aware methods, DLQ, publisher)
- **Tenant manager**: updated for v4 interfaces, LRU eviction, log compatibility layer
- **Crypto**: credential-safe `fmt` output, `String()`/`GoString()` redaction
- **Security**: sensitive field detection with normalized matching

### Documentation
- `AGENTS.md`: repository-specific guidance for coding agents
- `MIGRATION_MAP.md`: v1→v4 symbol mapping for consumers
- `docs/PROJECT_RULES.md`: full coding standards and architecture patterns
- `REVIEW.md`: review checklist and standards

### Testing
- Comprehensive unit tests across all packages
- Integration tests for Postgres, MongoDB, Redis, RabbitMQ, HTTP health, rate limiting, server lifecycle
- Benchmark tests for assertions and safe operations

---

**316 files changed** | ~70k lines added | ~14k lines removed

### CODE REVIEW

File: commons/app_test.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+func TestRunWithError(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil_logger_returns_ErrLoggerNil", func(t *testing.T) {
+		t.Parallel()
+
+		l := NewLauncher()
+		err := l.RunWithError()
+		assert.ErrorIs(t, err, ErrLoggerNil)
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/app_test.go]: The symbol is correctly named `ErrLoggerNil` -- it's defined at `commons/app.go:16` as `var ErrLoggerNil = errors.New("logger is nil")`. The test correctly references this symbol. No change needed.

File: commons/circuitbreaker/healthchecker_test.go [hunk 2]
```diff
@@ -67,63 +72,410 @@ func TestNewHealthCheckerWithValidation_NegativeTimeout(t *testing.T) {
...
+	// Fill the immediateCheck channel (capacity 10)
+	for i := 0; i < 10; i++ {
+		hcInternal.immediateCheck <- "fill"
+	}
+
+	// This should not block or panic — it logs a warning instead
+	assert.NotPanics(t, func() {
+		hcInternal.OnStateChange(context.Background(), "overflow-svc", StateClosed, StateOpen)
+	})
 }
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/circuitbreaker/healthchecker_test.go]: The channel capacity is an internal implementation detail of the health checker. Extracting it to a package-level constant would expose an implementation choice as API surface without adding clarity. The test intentionally mirrors the implementation to verify the specific buffering behavior.

File: commons/circuitbreaker/healthchecker.go [hunk 3]
```diff
@@ -58,45 +67,59 @@ func NewHealthCheckerWithValidation(manager Manager, interval, checkTimeout time
...
+
+	runtime.SafeGoWithContextAndComponent(
+		context.Background(),
+		hc.logger,
+		"circuitbreaker",
+		"health_check_loop",
+		runtime.KeepRunning,
+		func(ctx context.Context) {
+			hc.healthCheckLoop(ctx)
+		},
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/circuitbreaker/healthchecker.go]: Intentional. Health checks are infrastructure-level concerns that must continue running independently of any request context. `context.Background()` is the correct choice here -- the health loop's lifecycle is managed by `stopChan`, which `Stop()` closes. Using a request-scoped context would cause health checks to stop when unrelated requests complete.

File: commons/license/manager_test.go [hunk 4]
```diff
@@ -178,13 +140,49 @@ func TestTerminateSafe_UninitializedManager(t *testing.T) {
...
+	// WithFailClosed should set the handler to TerminateSafe behavior.
+	manager := license.New(license.WithFailClosed())
+
+	// TerminateSafe returns nil when handler is non-nil (it invokes handler then returns nil).
+	// The WithFailClosed handler itself calls TerminateSafe internally.
+	// Since the manager IS initialized (New was called), the handler should not return errors.
+	assert.NotPanics(t, func() {
+		manager.Terminate("fail-closed test")
+	})
 }
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/license/manager_test.go]: Not all tests in this file are parallelization-safe -- several modify global state (`SetHandler`, `Terminate`). Adding `t.Parallel()` to those would introduce race conditions. The tests that are parallel-safe already have it.

File: commons/license/manager.go [hunk 5]
```diff
@@ -83,27 +151,50 @@ func (m *ManagerShutdown) Terminate(reason string) {
...
+	}
+
+	if m.Logger != nil {
+		m.Logger.Log(context.Background(), log.LevelWarn, "license validation failed",
+			log.String("reason", reason),
+		)
+	}
+
 	return fmt.Errorf("%w: %s", ErrLicenseValidationFailed, reason)
 }
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/license/manager.go]: The warn-level log in `TerminateWithError` is intentional. License validation termination is a significant operational event regardless of the reason -- ops teams need visibility into when and why license checks trigger shutdown. Lowering to debug would hide this from standard monitoring pipelines.

File: commons/log/doc.go [hunk 6]
```diff
@@ -0,0 +1,5 @@
+// Package log defines the v2 logging interface and typed logging fields.
+//
+// Adapters (such as the zap package) implement Logger so applications can keep
+// logging calls consistent across backends.
+package log
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/log/doc.go]: "v2" refers to the internal API iteration of the logging interface, not the module version. The logging contract went through v1 (printf-style) and v2 (structured fields). This is documented in the comment itself. Changing it to "v4" would be misleading -- the logging API iteration is independent of the module version.

File: commons/circuitbreaker/manager.go [hunk 7]
```diff
@@ -202,57 +284,141 @@ func (m *manager) RegisterStateChangeListener(listener StateChangeListener) {
...
+	listener StateChangeListener,
+	serviceName string,
+	fromState State,
+	toState State,
+) {
+	listenerCtx, listenerCancel := context.WithTimeout(ctx, stateChangeListenerTimeout)
+	defer listenerCancel()
+
+	listener.OnStateChange(listenerCtx, serviceName, fromState, toState)
+}
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/circuitbreaker/manager.go]: The timeout context IS threaded through to the listener. `notifyStateChangeListener` creates `listenerCtx` with `context.WithTimeout(ctx, stateChangeListenerTimeout)` and passes it to `listener.OnStateChange(listenerCtx, ...)`. The listener is contractually obligated to honor the context -- this is the standard Go pattern for timeout propagation. Force-killing the goroutine would violate Go's cooperative cancellation model and risk resource leaks.

File: commons/assert/doc.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+//	assert.NonNegativeDecimal(d decimal.Decimal) bool // d >= 0
+//
+// Use predicates with Asserter:
+//
+//	if err := a.That(ctx, assert.Positive(count), "count must be positive", "count", count); err != nil {
+//	    return err
+//	}
+//	if err := a.That(ctx, assert.ValidUUID(id), "invalid UUID", "id", id); err != nil {
+//	    return err
+//	}
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/assert/doc.go]: Fixed.

File: commons/circuitbreaker/manager.go [hunk 9]
```diff
@@ -1,79 +1,168 @@
...
 
-// NewManager creates a new circuit breaker manager
-func NewManager(logger log.Logger) Manager {
-	return &manager{
+// NewManager creates a new circuit breaker manager.
+// Returns an error if logger is nil.
+func NewManager(logger log.Logger, opts ...ManagerOption) (Manager, error) {
+	if logger == nil {
+		return nil, ErrNilLogger
+	}
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/circuitbreaker/manager.go]: Fixed. Added `isNilLogger` helper mirroring the existing `isNilListener` pattern. `NewManager` now catches typed-nil loggers.

File: commons/assert/doc.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+//  1. Metrics: Records assertion_failed_total with component/operation/assertion labels.
+//     Initialize with InitAssertionMetrics(metricsFactory).
+//
+//  2. Tracing: Records assertion.failed span events (with stack traces in non-prod).
+//     Automatically uses the span from the context.
+//
+// # Stack Traces
+//
+// Stack traces are included in logs and trace events only in non-production
+// environments (ENV != production and GO_ENV != production).
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/assert/doc.go]: Fixed.

File: commons/mongo/mongo.go [hunk 11]
```diff
@@ -1,120 +1,732 @@
...
-	if err != nil {
-		mc.Logger.Errorf("failed to open connect to mongodb: %v", err)
-		return fmt.Errorf("failed to connect to mongodb: %w", err)
+	if cfg.TLS != nil && strings.TrimSpace(cfg.TLS.CACertBase64) == "" {
+		return configError("TLS CA cert is required when TLS is configured")
 	}
 
-	if err := noSQLDB.Ping(ctx, nil); err != nil {
-		mc.Logger.Errorf("MongoDBConnection.Ping failed: %v", err)
+	return nil
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/mongo/mongo.go]: Fixed.

File: commons/backoff/backoff_test.go [hunk 12]
```diff
@@ -0,0 +1,450 @@
...
+					assert.GreaterOrEqual(t, result, time.Duration(0))
+					assert.Less(t, result, maxDelay)
+				}
+			} else {
+				result := ExponentialWithJitter(tt.base, tt.attempt)
+				assert.Equal(t, tt.expected, result)
+			}
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/backoff/backoff_test.go]: Declining. With only 3 test cases, the current `expected == 0 && base > 0` branch is clear enough in context. Adding an `expectRange` flag for this size of table is over-engineering — the test struct stays simpler without the extra field, and the assertion logic is easy to follow. The important edge-case behavior is well-tested as-is.

File: commons/mongo/mongo.go [hunk 13]
```diff
@@ -1,120 +1,732 @@
...
+		clientOptions.SetMaxPoolSize(c.cfg.MaxPoolSize)
+	}
+
+	if c.cfg.TLS != nil {
+		tlsCfg, err := buildTLSConfig(*c.cfg.TLS)
 		if err != nil {
-			mc.Logger.Infof("ERRCONECT %s", err)
-			return nil, err
+			return fmt.Errorf("%w: TLS configuration: %w", ErrConnect, err)
+		}
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/mongo/mongo.go]: Fixed.

File: commons/assert/assert_extended_test.go [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
+
+	require.False(t, shouldIncludeStack())
+}
+
+func TestShouldIncludeStack_ProductionCaseInsensitive(t *testing.T) {
+	t.Setenv("ENV", "Production")
+	t.Setenv("GO_ENV", "")
+
+	require.False(t, shouldIncludeStack())
+}
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/assert/assert_extended_test.go]: **Implemented.** Added `// Not parallel - uses t.Setenv and depends on runtime global state.` comments to all four `shouldIncludeStack` tests that were missing them (`TestShouldIncludeStack_NonProduction`, `_ProductionENV`, `_ProductionGOENV`, `_ProductionCaseInsensitive`).

File: commons/mongo/mongo.go [hunk 15]
```diff
@@ -1,120 +1,756 @@
...
+func sanitizeDriverError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	msg := err.Error()
+	msg = uriCredentialsPattern.ReplaceAllString(msg, "://***@")
+	msg = uriPasswordParamPattern.ReplaceAllString(msg, "${1}***")
+
+	return &SanitizedError{Message: msg}
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply commons/mongo/mongo.go]: **Implemented.** Added `cause error` field and `Unwrap() error` method to `SanitizedError`. `sanitizeDriverError` now passes the original error as `cause`. This preserves the error chain so `errors.Is(err, context.Canceled)` / `context.DeadlineExceeded` / driver sentinels all work through the sanitization layer.

File: Makefile [hunk 16]
```diff
@@ -86,34 +163,319 @@ build:
...
+	$(call check_command,go,"Install Go from https://golang.org/doc/install")
+	$(call check_command,docker,"Install Docker from https://docs.docker.com/get-docker/")
+	@set -e; mkdir -p $(TEST_REPORTS_DIR); \
+	if [ -n "$(PKG)" ]; then \
+	  echo "Using specified package: $(PKG)"; \
+	  pkgs=$$(go list $(PKG) 2>/dev/null | tr '\n' ' '); \
+	else \
+	  echo "Finding packages with *_integration_test.go files..."; \
+	  dirs=$$(find . -name '*_integration_test.go' -not -path './vendor/*' 2>/dev/null | xargs -n1 dirname 2>/dev/null | sort -u | tr '\n' ' '); \
+	  pkgs=$$(if [ -n "$$dirs" ]; then go list $$dirs 2>/dev/null | tr '\n' ' '; fi); \
```
[the remark this answers is not in the document]
Comment (fredcamaral) [reply Makefile]: **Implemented.** Replaced `xargs -n1 dirname` with `find ... -exec dirname {} \;` in both `test-integration` (line 256) and `coverage-integration` (line 381). The `-exec` form handles empty results gracefully under `set -e` without the non-zero exit from `xargs`.
